services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java

```diff
@@ -0,0 +1,192 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.commons.lang.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This service is in charge of swapping to a new version after a specified wait time in the remaining regions of a target region push if enabled.
+ * The wait time is specified through a store/version level config (target_swap_region_wait_time) and the default wait time is 60m.
+ */
+public class DeferredVersionSwapService extends AbstractVeniceService {
+  private final AtomicBoolean stop = new AtomicBoolean(false);
+  private final Set<String> allClusters;
+  private final VeniceControllerMultiClusterConfig veniceControllerMultiClusterConfig;
+  private final VeniceParentHelixAdmin veniceParentHelixAdmin;
+  private final ScheduledExecutorService deferredVersionSwapExecutor = Executors.newSingleThreadScheduledExecutor();
+  private static final Logger LOGGER = LogManager.getLogger(DeferredVersionSwapService.class);
+
+  public DeferredVersionSwapService(
+      VeniceParentHelixAdmin admin,
+      VeniceControllerMultiClusterConfig multiClusterConfig) {
+    this.veniceParentHelixAdmin = admin;
+    this.allClusters = multiClusterConfig.getClusters();
+    this.veniceControllerMultiClusterConfig = multiClusterConfig;
+  }
+
+  @Override
+  public boolean startInner() throws Exception {
+    deferredVersionSwapExecutor.scheduleAtFixedRate(
+        new DeferredVersionSwapTask(),
+        0,
+        veniceControllerMultiClusterConfig.getDeferredVersionSwapSleepMs(),
+        TimeUnit.MILLISECONDS);
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    stop.set(true);
+    deferredVersionSwapExecutor.shutdown();
+  }
+
+  private Set<String> getRegionsForVersionSwap(Map<String, Integer> candidateRegions, Set<String> targetRegions) {
+    Set<String> remainingRegions = new HashSet<>(candidateRegions.keySet());
+    remainingRegions.removeAll(targetRegions);
+    return remainingRegions;
+  }
+
+  private String getTargetRegion(Set<String> targetRegions) {
+    return targetRegions.iterator().next();
+  }
+
+  private StoreResponse getStoreForRegion(String clusterName, Set<String> targetRegions, String storeName) {
+    Map<String, ControllerClient> controllerClientMap =
+        veniceParentHelixAdmin.getVeniceHelixAdmin().getControllerClientMap(clusterName);
+    ControllerClient targetRegionControllerClient = controllerClientMap.get(getTargetRegion(targetRegions));
+    return targetRegionControllerClient.getStore(storeName);
+  }
+
+  private class DeferredVersionSwapTask implements Runnable {
+    @Override
+    public void run() {
+      while (!stop.get()) {
+        for (String cluster: allClusters) {
```

Comment:
put a big try-catch block around this logic and add metric in the catch block to detect failed version swaps